### PR TITLE
feat: Support for async subscriber callbacks

### PR DESCRIPTION
**Pull request by ffMathy**
Add support for async subscriber callbacks while retaining backwards compatibility.

- [ ] I have added automatic tests where applicable
- [X] The PR title is suitable as a release note
- [X] The PR contains a description of what has been changed
- [X] The description contains manual test instructions

## Changes
Made it possible to have an async subscriber callback.

## Test
Check if subscribers are still called.

### CODE REVIEW

**Comment by gismya** [comment]
Looks great!

Will probably be a few weeks before this can be reviewed and merged. I won't have time to do it before going on summer vacations and the other maintainers are already on vacations.

**Review by gismya — CHANGES_REQUESTED** [review]
Looking good. Please add tests for handling both async and non async callbacks.

**Comment by ffMathy** [comment]
Okay. I looked at the existing tests, and I couldn't find enough inspiration to get started with this easily.

Do you have an idea of where I can start? The existing tests don't seem to have an example that tests the return value of the callback itself, and if that is being replied.

**Comment by gismya** [comment]
Something like this should work

```
  test("should handle async callback and return correct data", () => {
    const callback = vi.fn(async () => {
      return new Promise((resolve) => setTimeout(() => resolve("someData"), 1));
    });
    const testEvent = {
      topic: "ftrack.test",
      data: {},
      id: "eventId",
      source: { id: "sourceId" },
    };

    const publishReplySpy = vi
      .spyOn(eventHub, "publishReply")
      .mockImplementation((_, data) => data);

    eventHub.subscribe("topic=ftrack.test", callback);
    eventHub._handle(testEvent);

    setTimeout(() => {
      expect(callback).toHaveBeenCalledWith(testEvent);
      expect(publishReplySpy).toHaveBeenCalledWith(
        expect.anything(),
        "someData",
        expect.anything()
      );
    }, 10);
    publishReplySpy.mockRestore();
  });

  test("should handle sync callback and return correct data", () => {
    const callback = vi.fn(() => "someData");
    const testEvent = {
      topic: "ftrack.test",
      data: {},
      id: "eventId",
      source: { id: "sourceId" },
    };

    const publishReplySpy = vi
      .spyOn(eventHub, "publishReply")
      .mockImplementation((_, data) => data);

    eventHub.subscribe("topic=ftrack.test", callback);
    eventHub._handle(testEvent);

    setTimeout(() => {
      expect(callback).toHaveBeenCalledWith(testEvent);
      expect(publishReplySpy).toHaveBeenCalledWith(
        expect.anything(),
        "someData",
        expect.anything()
      );
    }, 10);
    publishReplySpy.mockRestore();
  });
});
```

Not the prettiest tests, but without changing the return values we have no way of awaiting everything, so there's a magic 10 ms timeout.

File: test/event_hub.test.js
Comment (gismya) [inline test/event_hub.test.js:210]:
```suggestion
  test("should handle sync callback and return correct data", async () => {
    const callback = vi.fn(() => "someData");
    const testEvent = {
      topic: "ftrack.test",
      data: {},
      id: "eventId",
      source: { id: "sourceId" },
    };

    const publishReplySpy = vi
      .spyOn(eventHub, "publishReply")
      .mockImplementation((_, data) => data);

    eventHub.subscribe("topic=ftrack.test", callback);
    const promises = eventHub._handle(testEvent);
    await Promise.all(promises);
    expect(callback).toHaveBeenCalledWith(testEvent);
    expect(publishReplySpy).toHaveBeenCalledWith(
      expect.anything(),
      "someData",
      expect.anything()
    );
    publishReplySpy.mockRestore();
  });

  test("should not handle async callback with a promise", async () => {
    const asyncCallback = vi.fn(async () => "someData");
    const testEvent = {
      topic: "ftrack.test",
      data: {},
      id: "eventId",
      source: { id: "sourceId" },
    };

    const publishReplySpy = vi
      .spyOn(eventHub, "publishReply")
      .mockImplementation((_, data) => data);

    eventHub.subscribe("topic=ftrack.test", asyncCallback);
    const promises = eventHub._handle(testEvent);
    await Promise.all(promises ?? []);
    expect(asyncCallback).toHaveBeenCalledWith(testEvent);
    expect(publishReplySpy).toHaveBeenCalled();
    expect(publishReplySpy).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.any(Promise),
      expect.anything()
    );
    expect(publishReplySpy).toHaveBeenCalledWith(
      expect.anything(),
      "someData",
      expect.anything()
    );
    publishReplySpy.mockRestore();
  });
```

File: source/event_hub.ts
Comment (gismya) [inline source/event_hub.ts:568]:
```suggestion
    const promises: Promise<any>[] = [];
```

File: source/event_hub.ts
Comment (gismya) [inline source/event_hub.ts:571]:
```suggestion
                promises.push(responsePromise);
```

File: source/event_hub.ts
Comment (gismya) [inline source/event_hub.ts:586]:
```suggestion
         return promises; 
    }
```

**Review by gismya — CHANGES_REQUESTED** [review]
Didn't actually test the test properly yesterday. It incorrectly passed even when a promise was sent as a response to publishReply, and didn't correctly check the value either. Adding the promises as return value for the function and awaiting (With some other changes) made the tests work as intended.

Good catch with .ts missing from lint-staged. I wonder why you need `npx` in the command though - could you check that you've done `npm install` (Or 'yarn' if that's what you're using) and that it's simply not missing the needed node modules? Otherwise I'll investigate if there might be any issues with using `npx` together with Yarn.

**Review by gismya — APPROVED** [review]
I did some mistakes when line by line suggesting the promise returns. I've commited the correct implementation.

You can remove the changes to package.json - we identified that the commit hooks have more issues and I'm posting a separate PR for that shortly. Thank you for pointing that out! Our IDE setups made us miss that it didn't run correctly.

**Comment by ffMathy** [comment]
Thanks for this!

**Pull request merged**
